code/game/objects/structures/curtains.dm

```diff
@@ -48,16 +149,27 @@
 	opacity = TRUE
 	density = FALSE
 	anchored = TRUE
-
-	var/obj/item/curtain/holder = /obj/item/curtain
+	var/curtain_kind_path = /decl/curtain_kind
 
 /obj/structure/curtain/open
 	icon_state = "open"
 	layer = ABOVE_HUMAN_LAYER
 	opacity = FALSE
 
-/obj/structure/curtain/Initialize()
+/obj/structure/curtain/Initialize(ml, _mat, _reinf_mat, var/decl/curtain_kind/kind = null)
+	if(kind)
+		if(istype(kind))
+			curtain_kind_path = kind.type
+		else if(ispath(kind))
+			curtain_kind_path = kind
+	else
+		kind = GET_DECL(curtain_kind_path)
+	_mat = kind.material_key
+	matter = atom_info_repository.get_matter_for(/obj/structure/curtain, kind.material_key)
 	. = ..()
+	src.SetName(kind.name)
+	src.alpha = kind.alpha
```

Comment:
As with the above comment this should probably be moved into an icon update or update_materials override.